### PR TITLE
 [AArch64][GlobalISel] Combine MUL(AND(LSHR(X, 15), 0x10001), 0xffff) to CMLTz

### PR DESCRIPTION
**Pull request by chuongg3**

### CODE REVIEW

File: llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp [hunk 1]
```diff
@@ -381,6 +381,60 @@ void applyOrToBSP(MachineInstr &MI, MachineRegisterInfo &MRI,
   MI.eraseFromParent();
 }
 
+bool matchCombineMulCMLT(MachineInstr &MI, MachineRegisterInfo &MRI,
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp:384]: Add a comment, similar to the pattern.

File: llvm/test/CodeGen/AArch64/mulcmle.ll [hunk 2]
```diff
@@ -1,11 +1,22 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
-; RUN: llc -mtriple=aarch64 %s -o - | FileCheck %s
+; RUN: llc -mtriple=aarch64 %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-SD
+; RUN: llc -mtriple=aarch64 %s -o - -global-isel -global-isel-abort=2 2>&1 | FileCheck %s --check-prefixes=CHECK,CHECK-GI
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/mulcmle.ll:3]: This might be able to remove `-global-isel-abort=2 2>&1` now that all the tests do not fall back.

**Review by davemgreen — COMMENTED** [review]
It might be good to say this is the equivalent of https://reviews.llvm.org/D130874 in the commit message.

File: llvm/test/CodeGen/AArch64/mulcmle.ll [hunk 3]
```diff
@@ -1,11 +1,22 @@
...
+;
+; CHECK-GI-LABEL: v1i64:
+; CHECK-GI:       // %bb.0:
+; CHECK-GI-NEXT:    fmov x8, d0
+; CHECK-GI-NEXT:    lsr x8, x8, #31
+; CHECK-GI-NEXT:    and x8, x8, #0x100000001
+; CHECK-GI-NEXT:    lsl x9, x8, #32
+; CHECK-GI-NEXT:    sub x8, x9, x8
+; CHECK-GI-NEXT:    fmov d0, x8
+; CHECK-GI-NEXT:    ret
```
Comment (madhur13490) [inline llvm/test/CodeGen/AArch64/mulcmle.ll:19]: Are you planning to optimize this sequence in some further patch? I see this is sub-optimal compared to SDAG.

Comment (chuongg3) [reply llvm/test/CodeGen/AArch64/mulcmle.ll:19]: In GlobalISel, all v1 types are treated as scalar values and we do not necessarily want the same optimization for scalar types.

**Review by davemgreen — APPROVED** [review]
LGTM, thanks

**Pull request merged**